src/enumerable.cr

```diff
@@ -1634,7 +1634,7 @@ module Enumerable(T)
   # a.zip(b, c) # => [{1, 4, 8}, {2, 5, 7}, {3, 6, 6}]
   # ```
   def zip(*others : Indexable | Iterable | Iterator)
-    pairs = Array(typeof(zip(*others) { |e| break e }.not_nil!)).new(size)
+    pairs = Array(typeof(zip(*others) { |e| break e }.not_nil!)).new
```

Comment:
I mean if `other` is an `Indexable`, we could use its `size`?

Reply:
That's not correct, the size is always taken from `self`.

Reply:
Speficially, `[1, 2, 3].zip(4..5)` should give `[[1, 4], [2, 5]]`. Even if we take the size from the other indexable, it might not be enough and there will be a reallocation.

In short, maybe it can be done, but I don't know. Better to be safe.

Reply:
@asterite: `[1,2,3].zip(4..5)` raises an `IndexError` as written in the documentation. All enumerable must have at least as many elements as `self`. So the size of the returned Array is always *exactly* `self.size` or an exception is thrown.

Reply:
I think it'd be nice to keep this optimization in for `Indexable`. `self.is_a?(Indexable) ? size : 0` should work well.

Reply:
@RX14: thanks. I did exactly that!